prow/githubeventserver/githubeventserver.go

```diff
@@ -0,0 +1,495 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package githubeventserver
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/pluginhelp"
+	pluginhelp_externalplugins "k8s.io/test-infra/prow/pluginhelp/externalplugins"
+	pluginhelp_hook "k8s.io/test-infra/prow/pluginhelp/hook"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const (
+	eventTypeField = "event-type"
+
+	statusEvent                   = "status"
+	pushEvent                     = "push"
+	pullRequestReviewCommentEvent = "pull_request_review_comment"
+	pullRequestReviewEvent        = "pull_request_review"
+	pullRequestEvent              = "pull_request"
+	issueCommentEvent             = "issue_comment"
+	issuesEvent                   = "issues"
+)
+
+// GitHubEventServer hold all the information needed for the
+// github event server implementation.
+type GitHubEventServer struct {
+	wg *sync.WaitGroup
+
+	endpoint string
+	port     int
+
+	serveMuxHandler *serveMuxHandler
+	httpServeMux    *http.ServeMux
+}
+
+// NewGitHubEventServer creates a new *GitHubEventServer from the given arguments.
+// It also assigns the serveMuxHandler in the http.ServeMux.
+func NewGitHubEventServer(hmacTokenGenerator func() []byte, options Options, metrics *Metrics, log *logrus.Entry) *GitHubEventServer {
```

Comment:
I think the options should contain all possible options

Reply:
Well, the server only needs the default endpoint and port. What other possible options you mean?

Reply:
`hmacTokenGenerator`, `metrics`, `log`. It is not logical to have an options struct that holds two out of five mandatory options

Reply:
Yea I guess, the consumer can just update the options after flags passed.

Reply:
Also, I didn't want to have unnecessary exposed fields in Options struct. So I think passing them in the `NewGitHubEventServer` without messing around with the Options, should be sufficient. WDYT?

Reply:
No. Either we don't have no options struct at all or it holds at all options. A middle ground is just super confusing.

Reply:
Makes sense. Let me address this.

Reply:
Addressed. Maybe it's better to add some helper functions to `Set/Get` instead of exposing them.